sdk/src/trace/batch_span_processor.cc

```diff
@@ -0,0 +1,184 @@
+#pragma once
+
+#include "opentelemetry/sdk/trace/batch_span_processor.h"
+#include "src/common/circular_buffer.h"
+
+#include <vector>
+
+using opentelemetry::sdk::common::AtomicUniquePtr;
+using opentelemetry::sdk::common::CircularBuffer;
+using opentelemetry::sdk::common::CircularBufferRange;
+
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace trace
+{   
+
+BatchSpanProcessor::BatchSpanProcessor(std::unique_ptr<SpanExporter>&& exporter,
+                                       const int max_queue_size,
+                                       const int schedule_delay_millis,
+                                       const int max_export_batch_size)
+       :exporter_(std::move(exporter)),
+        max_queue_size_(max_queue_size), 
+        schedule_delay_millis_(schedule_delay_millis),
+        max_export_batch_size_(max_export_batch_size)   
+{  
+    buffer_ = std::unique_ptr<CircularBuffer<Recordable>>(new CircularBuffer<Recordable>(max_queue_size_));
+    
+    //Start the background worker thread
+    worker_thread_ = std::unique_ptr<std::thread>(new std::thread(&BatchSpanProcessor::DoBackgroundWork, this));
+
+} 
+
+std::unique_ptr<Recordable> BatchSpanProcessor::MakeRecordable() noexcept 
+{
+    return exporter_->MakeRecordable();
+}
+
+void BatchSpanProcessor::OnStart(Recordable &span) noexcept  
+{
+    // no-op
+}
+
+void BatchSpanProcessor::OnEnd(std::unique_ptr<Recordable> &&span) noexcept 
+{
+    if(is_shutdown_){
+        // TODO: log in glog
+        return;
+    }
+
+    if(static_cast<int>(buffer_->size()) == max_queue_size_){
+        // TODO: glog that spans will likely be dropped
+    }
+
+    buffer_->Add(span);
+
+    // If the queue gets half full a preemptive notification is sent to the worker thread to
+    // start a new export cycle.
+    if(static_cast<int>(buffer_->size()) == max_queue_size_ / 2){
+        // signal the worker thread
+        cv_.notify_one();
+    }
+}
+
+void BatchSpanProcessor::ForceFlush(std::chrono::microseconds timeout) noexcept 
+{    
+    if(is_shutdown_ == true){
+        // TODO: glog that processor is already shutdown
+    }
+
+    is_force_flush_ = true;
+    buffer_->Add(FORCE_FLUSH_TOKEN);
+
+    // wake up the worker thread
+    cv_.notify_one();
```

Comment:
There is a high possibility that this notification call is missed because of lack of synchronization.

Reply:
A safer way to do this would be to verify that notification was not missed by looping over it. You can use 
is_force_flush_ for that, something like that:

std::chrono::microseconds timeout_left = timeout;
while(is_force_flush_ && timeout_left >0) {
  auto start = std::chrono::steady_clock::now();
  cv_.notify_one();
  std::this_thread::sleep_for(std::chrono::nanoseconds(50));
  auto end = std::chrono::steady_clock::now();
  timeout_left -= (end-start);
}

and then set is_force_flush_ to false in the DoBackgroundWork if it was set to true when it wakes up. That way there is no need for a token, force_flush_lk and force_flush_cv_ - this thread will keep trying to wake up worker until timeout expires or until is_force_flush_ becomes true.

Reply:
@reyang There was a small mistake in this message I sent on gitter channel



> I think one way to avoid all of this would be to make the condition variable notify() more robust, as mentioned in [this comment](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/sdk/include/opentelemetry/sdk/trace/exporter.h#L15)


The link was incorrect. Apparently, the 'Copy Link' from the drop-down menu of the comment is not actually getting copied to clipboard for some reason.

Anyways, I was referring to the code above. Also, your implementation works in [OpenCensus Python](https://github.com/census-instrumentation/opencensus-python/blob/b7c7a67d7bcae749c19a3508777522f206da75cf/contrib/opencensus-ext-azure/opencensus/ext/azure/common/exporter.py#L84) works too!